internal/server/authz/engine.go

```diff
@@ -2,105 +2,179 @@ package authz
 }
 
-type Engine struct {
-	query rego.PreparedEvalQuery
+type CachedSource[T any] interface {
+	Get(_ context.Context, hash []byte) (T, []byte, error)
 }
 
-var defaultPolicy = `package authz.v1
-import data
+type PolicySource CachedSource[[]byte]
 
-default allow = false
+type DataSource CachedSource[map[string]any]
 
-allow {
-    input.role != ""
-    input.action != ""
-    input.scope != ""
+type Engine struct {
+	logger *zap.Logger
 
-    permissions := get_permissions(input.role)
-    allowed(permissions, input.action, input.scope)
+	mu    sync.RWMutex
+	query rego.PreparedEvalQuery
+	store storage.Store
+
+	policySource PolicySource
+	policyHash   []byte
+
+	dataSource DataSource
+	dataHash   []byte
+
+	pollDuration time.Duration
 }
 
-get_permissions(role) = result {
-    some idx
-    data.roles[idx].name = role
-    result = data.roles[idx].rules
+func WithDataSource(source DataSource) containers.Option[Engine] {
+	return func(e *Engine) {
+		e.dataSource = source
+	}
 }
 
-allowed(permissions, action, scope) {
-    permissions[action]  # First, ensure the action key exists
-    scope_in_list(permissions[action], scope)  # Check if the scope is in the list
+func WithPollDuration(dur time.Duration) containers.Option[Engine] {
+	return func(e *Engine) {
+		e.pollDuration = dur
+	}
 }
 
-allowed(permissions, action, scope) {
-    permissions[action]["*"]  # Checks if all scopes are allowed for the action
+func NewEngine(ctx context.Context, logger *zap.Logger, source PolicySource, opts ...containers.Option[Engine]) (*Engine, error) {
+	engine := &Engine{
+		logger:       logger,
+		policySource: source,
+		store:        inmem.New(),
+		pollDuration: defaultPollDuration,
+	}
+
+	containers.ApplyAll(engine, opts...)
+
+	err := engine.update(ctx, storage.AddOp)
+	if err != nil {
+		return nil, err
+	}
+
+	go engine.pollData(ctx)
+
+	return engine, nil
 }
 
-# Handles cases where "*" is provided for all actions
-allowed(permissions, action, scope) {
-    permissions["*"] != null  # Check if wildcard for all actions exists
-    scope_in_list(permissions["*"], scope)  # Check if scope is universally allowed or specific to an action
+func (e *Engine) IsAllowed(ctx context.Context, input map[string]interface{}) (bool, error) {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+
+	results, err := e.query.Eval(ctx, rego.EvalInput(input))
+	if err != nil {
+		return false, err
+	}
+
+	if len(results) == 0 {
+		return false, nil
+	}
+
+	return results[0].Expressions[0].Value.(bool), nil
 }
 
-# Helper to handle array membership or wildcard
-scope_in_list(list, scope) {
-    list[_] = scope  # Scope is explicitly listed in the permissions
+func (e *Engine) pollData(ctx context.Context) {
+	ticker := time.NewTicker(e.pollDuration)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if err := e.update(ctx, storage.ReplaceOp); err != nil {
+				e.logger.Error("updating policy and data", zap.Error(err))
```

Comment:
Just to double check, we want to update both policy and data on the same cadence?

Reply:
yeah i mean i guess it doesnt matter too much since its a noop of the file didn't change, other than checking the modtime. im wondering which file would realistically change more often? likely the data, but i think i agree that simple is better from a config aspect so im down to keep it as is. just wanted to double check